# CE888 Lab9
### GAN generator netowrk
**Date**: 11 March 2019

In [1]:
import keras
from keras import layers
import numpy as np

Using TensorFlow backend.


In [2]:
parameters={0:{"gen_act":"tanh", "cls_act":"sigmoid",\
               "disc_loss":"binary_crossentropy"},\
            1:{"gen_act":"sigmoid", "cls_act":"sigmoid", \
               "disc_loss":"binary_crossentropy"},\
            2:{"gen_act":"tanh", "cls_act":"tanh", \
               "disc_loss":"binary_crossentropy"},\
            3:{"gen_act":"tanh", "cls_act":"sigmoid", \
               "disc_loss":"mean_squared_error"}}

para_id=int(input("Select parameters from [%s]: " % \
              ",".join([str(i) for i in list(parameters.keys())])))

Select parameters from [0,1,2,3]: 3


### GAN generator network

In [3]:
latent_dim = 32
height = 32
width = 32
channels = 3

generator_input = keras.Input(shape=(latent_dim,))

# Transforms the input into a 16 × 16 128-channel feature map
x = layers.Dense(128 * 16 * 16)(generator_input)
x = layers.LeakyReLU()(x)
x = layers.Reshape((16, 16, 128))(x)

x = layers.Conv2D(256, 5, padding='same')(x)
x = layers.LeakyReLU()(x)

# Upsamples to 32 × 32
x = layers.Conv2DTranspose(256, 4, strides=2, padding='same')(x)
x = layers.LeakyReLU()(x)

x = layers.Conv2D(256, 5, padding='same')(x)
x = layers.LeakyReLU()(x)
x = layers.Conv2D(256, 5, padding='same')(x)
x = layers.LeakyReLU()(x)

# Produces a 32 × 32 1-channel feature map (shape of a CIFAR10 image)
x = layers.Conv2D(channels, 7, activation=parameters[para_id]["gen_act"], \
                  padding='same')(x)

# Instantiates the generator model, which maps the input of shape (latent_dim,) 
#into an image of shape (32, 32, 3)
generator = keras.models.Model(generator_input, x)
generator.summary()

Instructions for updating:
Colocations handled automatically by placer.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 32)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 32768)             1081344   
_________________________________________________________________
leaky_re_lu_1 (LeakyReLU)    (None, 32768)             0         
_________________________________________________________________
reshape_1 (Reshape)          (None, 16, 16, 128)       0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 16, 16, 256)       819456    
_________________________________________________________________
leaky_re_lu_2 (LeakyReLU)    (None, 16, 16, 256)       0         
_________________________________________________________________
conv

### The GAN discriminator network

In [4]:
discriminator_input = layers.Input(shape=(height, width, channels))

x = layers.Conv2D(128, 3)(discriminator_input)
x = layers.LeakyReLU()(x)
x = layers.Conv2D(128, 4, strides=2)(x)
x = layers.LeakyReLU()(x)
x = layers.Conv2D(128, 4, strides=2)(x)
x = layers.LeakyReLU()(x)
x = layers.Conv2D(128, 4, strides=2)(x)
x = layers.LeakyReLU()(x)
x = layers.Flatten()(x)

# One dropout layer: an important trick
x = layers.Dropout(0.4)(x)

# Classification layer
x = layers.Dense(1, activation=parameters[para_id]["cls_act"])(x)

# Instantiates the discriminator model, which turns a (32, 32, 3) input into 
# a binary classification decision (fake/real)
discriminator = keras.models.Model(discriminator_input, x)
discriminator.summary()

# Uses gradient clipping (by value) in the optimizer and To stabilize training, 
# uses learning-rate decay
discriminator_optimizer = keras.optimizers.RMSprop(lr=0.0008,clipvalue=1.0,\
                                                   decay=1e-8)
discriminator.compile(optimizer=discriminator_optimizer,\
                      loss=parameters[para_id]["disc_loss"])

Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 32, 32, 3)         0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 30, 30, 128)       3584      
_________________________________________________________________
leaky_re_lu_6 (LeakyReLU)    (None, 30, 30, 128)       0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 14, 14, 128)       262272    
_________________________________________________________________
leaky_re_lu_7 (LeakyReLU)    (None, 14, 14, 128)       0         
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 6, 6, 128)         262272    
___________________________

### Adversarial network

In [0]:
# Sets discriminator weights to non-trainable
# (this will only apply to the gan model)
discriminator.trainable = False

gan_input = keras.Input(shape=(latent_dim,))
gan_output = discriminator(generator(gan_input))
gan = keras.models.Model(gan_input, gan_output)

gan_optimizer = keras.optimizers.RMSprop(lr=0.0004, clipvalue=1.0, decay=1e-8)
gan.compile(optimizer=gan_optimizer, loss='binary_crossentropy')

In [6]:
# Load data
from keras.preprocessing import image

# Loads CIFAR10 data
(x_train, y_train), (_, _) = keras.datasets.cifar10.load_data()

# Selects frog images (class 6)
x_train = x_train[y_train.flatten() == 6]

170500096/170498071 [==============================] - 111s 1us/step


In [7]:
print("x_train\t:",x_train.shape)
print("y_train\t:",y_train.shape)

x_train	: (5000, 32, 32, 3)
y_train	: (50000, 1)


In [0]:
# train the netowrk
x_train = x_train.reshape(
    (x_train.shape[0],) +
    (height, width, channels)).astype('float32') / 255.

In [9]:
# create output picture folder
import os

save_dir = "output_files"
if os.path.exists(save_dir)==False:
  os.mkdir(save_dir)
  print("<%s> directory is created." % save_dir)
else:
  print("<%s> directory already exists." % save_dir)

<output_files> directory is created.


In [10]:
# set parameters for random sampling
iterations = 10000
batch_size = 20

start = 0
# Samples random points in the latent space
for step in range(iterations):
    random_latent_vectors = np.random.normal(size=(batch_size, latent_dim))

# Decodes them to fake images    
generated_images = generator.predict(random_latent_vectors)

# Combines them with real images 
stop = start + batch_size
real_images = x_train[start: stop]
combined_images = np.concatenate([generated_images, real_images])

labels = np.concatenate([np.ones((batch_size, 1)), np.zeros((batch_size, 1))])
labels += 0.05 * np.random.random(labels.shape)

# Trains the discriminator
d_loss = discriminator.train_on_batch(combined_images, labels)

# Samples random points in the latent space
random_latent_vectors = np.random.normal(size=(batch_size,latent_dim))

# Assembles latent space labels that say “these are all real images”
# (it’s a lie!)
misleading_targets = np.zeros((batch_size, 1))

# Trains the generator (via the gan model, where the discriminator weights
# are frozen)
a_loss = gan.train_on_batch(random_latent_vectors, misleading_targets)

start += batch_size
if start > len(x_train) - batch_size: 
    start = 0
    
if step % 100 == 0: 
    gan.save_weights('gan.h5')
    
print('Discriminator loss:', d_loss)
print('Adversarial loss:', a_loss)

Instructions for updating:
Use tf.cast instead.


/usr/local/lib/python3.6/dist-packages/keras/engine/training.py:490: UserWarning: Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
  'Discrepancy between trainable weights and collected trainable'


Discriminator loss: 0.25884765
Adversarial loss: 0.26298517


In [0]:
# create and store output pictures
img = image.array_to_img(generated_images[0] * 255., scale=False)
img.save(save_dir + "/generated_frog_%s_%d.png" % (str(step), para_id))
         
img = image.array_to_img(real_images[0] * 255., scale=False)
img.save(save_dir+ "/real_frog_%s_%d.png" % (str(step), para_id))

### Results with different parameters:

|#|GAN generator activation function|Classification layer activation function|Discriminator loss functio|discriminator loss|adversarial loss|
|--------| --------|--------|--------|--------|--------|
|00|tanh|sigmoid|binary_crossentropy|0.7097865|0.29890946|
|01|sigmoid|sigmoid|binary_crossentropy|0.69281924|3.2741482e-07|
|02|tanh|tanh|binary_crossentropy|8.457808|0.6745621|
|03|tanh|sigmoid|mean_squared_error|0.25884765|0.26298517|